### ISSUE TITLE
WebGPU: Missing enumeration(s) in WGPUBlendFactor compared to Dawn

### ISSUE THREAD

**Issue by birsoyo** [emscripten-core/emscripten#21552]
I am using emscripten 3.1.55 for web and Dawn@ad853f80 for using webgpu natively on windows.

```
clang version 19.0.0git (https:/github.com/llvm/llvm-project 6c7805d5d186a6d1263f90b8033ad85e2d2633d7)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: [snip]\upstream\bin
```

As far as I can tell Dawn@ad853f80 should match emscripten 3.1.55 version of the webgpu headers.
 
When compiling with emscripten, the code fails with 
```
[snip]: error: no member named 'OneMinusSrc1' in 'wgpu::BlendFactor'; did you mean 'OneMinusSrc'?
  457 |       return wgpu::BlendFactor::OneMinusSrc1;
```

It looks like the following blend factor enumerations are missing.
```cpp
    WGPUBlendFactor_Src1 = 0x0000000E,
    WGPUBlendFactor_OneMinusSrc1 = 0x0000000F,
    WGPUBlendFactor_Src1Alpha = 0x00000010,
    WGPUBlendFactor_OneMinusSrc1Alpha = 0x00000011,
// Above are from Dawn@ad853f80
```

Are these missing enumerations intentional for not being supported on web?

Is there a place I can take a look how emscripten generates webgpu headers (from Dawn) so that I can match emscripten headers better with my native webgpu code?

**Comment by kainino0x** [comment]
Dual source blending hasn't landed in the WebGPU spec yet: https://github.com/gpuweb/gpuweb/issues/4283
More importantly you wouldn't be able to use it: These enums haven't been wired up in Chromium yet and I don't think we have implemented the shader compiler part for the latest spec decision. (I don't suspect any other browser would have implemented it yet either.) Filed https://crbug.com/330066244 about it.

**Comment by beaufortfrancois** [comment]
@kainino0x  Dual source blending is going to ship soon: https://groups.google.com/a/chromium.org/g/blink-dev/c/C3VH1EMrs1Y
1. Shall we update Emscripten?
2. Is there anything we should do in [emdawnwebgpu](https://dawn.googlesource.com/dawn/+/refs/heads/main/third_party/emdawnwebgpu/) as well? My naive understanding is that it's working already as https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/third_party/emdawnwebgpu/library_webgpu.js doesn't have any enum and feature names.

**Comment by kainino0x** [comment]
Basically the WebGPU bindings in Emscripten are a snapshot / in "maintenance mode" for the moment. The header currently in here was generated out of an older revision of Dawn, but even at that time there were differences between Dawn and Emscripten because Dawn has experimental features not available on the Web (including dual source blending at that time, before it was standardized).

Right now we are working on bringing the Emscripten bindings up to speed in a fork maintained inside Dawn instead of trying to do everything across two projects at once. So we strongly prefer to spend time on that fork so we can get it completed and merged back into Emscripten sooner rather than later.

@birsoyo do you need to use dual source blending for your application? (Or did you just encounter the difference in some other way?) If you wish to use Dawn's fork instead of Emscripten's snapshot you [can do that](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/emdawnwebgpu/README.md), but note that it is going to be receiving many small breaking changes over the next few months. Or if you want to contribute a small patch to add just this feature to Emscripten, please open a PR and @ me for review!

**Comment by kainino0x** [comment]
> 2\. Is there anything we should do in [emdawnwebgpu](https://dawn.googlesource.com/dawn/+/refs/heads/main/third_party/emdawnwebgpu/) as well? My naive understanding is that it's working already as https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/third_party/emdawnwebgpu/library_webgpu.js doesn't have any enum and feature names.

Correct, we shouldn't need to do anything, because emdawnwebgpu got rid of all the manually maintained tables in favor of autogeneration.

**Comment by beaufortfrancois** [comment]
@kainino0x Thank you for the detailed explanation! May I suggest updating https://github.com/emscripten-core/emscripten/blob/main/system/include/webgpu/README.md so that developers interested in latest changes know what to do?
Moreover, the README currently points to files in https://source.chromium.org/chromium/chromium/src/+/main:out/linux-Debug/gen/third_party/dawn/src/emdawnwebgpu/include which does not exist. Is this outdated or simply wrong?

**Comment by kainino0x** [comment]
> Moreover, the README currently points to files in https://source.chromium.org/chromium/chromium/src/+/main:out/linux-Debug/gen/third_party/dawn/src/emdawnwebgpu/include which does not exist. Is this outdated or simply wrong?

The link was correct but those generated files aren't getting indexed by code search anymore. IIRC might be because we moved some of that codegen stuff into the CMake build instead (which doesn't run for chromium code search). @lokokung would remember.

**Comment by kainino0x** [comment]
This won't be fixed in Emscripten's current implementation, see #23432 for migration to "emdawnwebgpu".
